keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,445 @@
+  * The node is shutdown
+  * The node is in other non-recoverable cases:
+    * hardware failure
+    * broken OS
+
+### Goals
+
+* Increase the availability of stateful workloads when a node is shutdown, or in other non-recoverable cases such as the hardware failure or broken OS.
+* Automate self-healing for stateful workloads.
+
+### Non-Goals
+
+* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
+  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
+* Implement in-cluster logic to handle node/control plane partitioning.
+
+## Proposal
+
+In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.
+
+Existing logic:
+1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+1. [Proposed change] This proposal requires a user to apply a `quarantine` taint on a node when the user has confirmed that this node is shutdown or in a non-recoerable state due to the hardware failure or broken OS.
+
+1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:
+  1.  Upon seeing the `quarantine` taint, the Pod GC Controller will forcefully delete the pods on the node. This should happen before the pods are being evicted. The `quarantine` taint indicates that new pods should not be scheduled on the node.
+  1. new `quarantine` taint has NoSchedule effect, meaning it will not schedule on the shutdown node. For example, `quarantine=nodeshutdown:NoSchedule` or `quarantine=hardwarefailure:NoSchedule`.
+
+1. [Proposed change] Once pods are selected and forcefully deleted, the attachdetach reconciler should check the `quarantine` taint on the node. If the taint is present, the attachdetach reconciler will not wait for 6 minutes to do force detach. Instead it will force detach right away and allow `volumeAttachement` to be deleted.
+
```

Comment:
attachdetach reconciler itself do force detach? If so, how to do?

Reply:
actually there is no actual difference between a normal detach operation and a "force detach". We call it force here mainly because normally controller will not trigger detach at all if node status showing volume is still in use (means volume is still mounted) by kubelet. But here we are sure that node is in shut down mode, we can trigger detach, no matter whether node status showing volume is in use or not..

Reply:
@jingxu97 Thank you, I understand :)